Add `CRUD` endpoints for todo app

devrajneupane opened the pull request:
This PR adds following endpoints

`GET` `/` Get all task
`GET` `/:id` Get task by id
`POST` `/` Create task
`PUT` `/:id` Update task
`DELETE` `/:id` Delete task

devrajneupane committed:
chore: add `.env` example

devrajneupane committed:
feat: setup project

devrajneupane committed:
feat: add `tasks` route

devrajneupane committed:
feat: add task controller

devrajneupane committed:
feat: add `Task` interface

devrajneupane committed:
feat: add task service

devrajneupane committed:
feat: add `TaskStatus` enum

devrajneupane committed:
feat: add task model

devrajneupane committed:
feat: setup docker for development

aman007shrestha commented on src/app.ts:
```diff
@@ -0,0 +1,22 @@
+    message: "Todo App",
+  });
+});
+
```
app.get(errorHandler) is the last piece of middleware =>  Parses App failures in response json format

{success: false,
message: "...",
statusCode: 500
}

aman007shrestha commented on src/service/taskService.ts:
```diff
@@ -0,0 +1,43 @@
+  const data = TaskModel.getTaskById(id);
+
+  if (!data) {
+    return {
```
throw Error from here and catch on controller. throw Error("Data Not Found")
advanced stage NotFoundError("....")

aman007shrestha commented on src/model/taskModel.ts:
```diff
@@ -0,0 +1,68 @@
+import { ITask } from "../interfaces/Task";
+import { TASK_STATUS } from "../enums/TaskStatus";
+
+const tasks = [
```
add Type for Task Object

aman007shrestha commented on src/controller/taskController.ts:
```diff
@@ -0,0 +1,52 @@
+
+export function createTask(req: Request, res: Response) {
+  const { body } = req;
+  TaskService.createTask(body);
```
Even if Service and Model fails to create task. The response would still be Task Created

aman007shrestha commented on src/model/taskModel.ts:
```diff
@@ -0,0 +1,68 @@
+  return tasks.find(({ id: userId }) => userId === id);
+}
+
+export function createTask(task: ITask) {
```
Service is expecting something right ?

aman007shrestha commented on src/service/taskService.ts:
```diff
@@ -0,0 +1,43 @@
+  const index = TaskModel.updateTask(id, task);
+
+  if (index === -1) {
+    return {
```
Throw and catch error

aman007shrestha commented on src/interfaces/Task.ts:
```diff
@@ -0,0 +1,7 @@
+export interface ITask {
+  title: string;
+  desc: string;
+  status: TASK_STATUS
```
union operator  "not started"| "pending" | "done"

devrajneupane committed:
chore: update heartbeat endpoint

devrajneupane committed:
chore: add sample `tasks.json`

devrajneupane committed:
feat: add utils

Utility functions to read and write `tasks.json`

devrajneupane committed:
chore: update task model

Update task model for better error handling

devrajneupane committed:
feat: update task service for better error handling

Appropriate response with custom message is send back to controller for
more convenient error handling

devrajneupane committed:
feat: add appropriate status code to response

devrajneupane committed:
refactor: minor code changes

devrajneupane committed:
feat: add service response interface

devrajneupane committed:
chore: add README.md
